core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
 
-    int versionHistorySizeToKeep =
-        PropertyUtil.propertyAsInt(
-            properties(),
-            ViewProperties.VERSION_HISTORY_SIZE,
-            ViewProperties.VERSION_HISTORY_SIZE_DEFAULT);
+      this.location = newLocation;
+      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
+      return this;
+    }
+
+    public Builder setCurrentVersionId(int newVersionId) {
+      if (newVersionId == -1) {
+        ValidationException.check(
+            lastAddedVersionId != null,
+            "Cannot set last version id: no current version id has been set");
+        return setCurrentVersionId(lastAddedVersionId);
+      }
+
+      if (this.currentVersionId == newVersionId) {
+        return this;
+      }
+
+      checkCurrentVersionIdIsValid(newVersionId);
+
+      // by that time, schemas are defined and the highestFieldId can be determined and schema
+      // changes can be added
+      int highestFieldId = highestFieldId();
+      for (Schema schema : schemas) {
+        this.changes.add(new MetadataUpdate.AddSchema(schema, highestFieldId));
+      }
+
+      this.currentVersionId = newVersionId;
 
-    if (versionHistorySizeToKeep <= 0) {
-      LOG.warn(
-          "{} must be positive but was {}",
+      if (lastAddedVersionId != null && lastAddedVersionId == newVersionId) {
+        this.changes.add(new MetadataUpdate.SetCurrentViewVersion(LAST_ADDED));
+      } else {
+        this.changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
+      }
+
+      return this;
+    }
+
+    private int highestFieldId() {
+      return schemas.stream().map(Schema::highestFieldId).max(Integer::compareTo).orElse(0);
+    }
+
+    public Builder setCurrentVersion(ViewVersion version) {
+      Schema schema = indexSchemas(schemas).get(version.schemaId());
+      int newSchemaId = addSchemaInternal(schema);
+      return setCurrentVersionId(
+          addVersionInternal(
+              ImmutableViewVersion.builder().from(version).schemaId(newSchemaId).build()));
+    }
+
+    private int reuseOrCreateNewViewVersionId(ViewVersion viewVersion) {
+      // if the view version already exists, use its id; otherwise use the highest id + 1
+      int newVersionId = viewVersion.versionId();
+      for (ViewVersion version : versions) {
+        if (version.equals(viewVersion)) {
+          return version.versionId();
+        } else if (version.versionId() >= newVersionId) {
+          newVersionId = viewVersion.versionId() + 1;
+        }
+      }
+
+      return newVersionId;
+    }
+
+    private int addVersionInternal(ViewVersion version) {
+      int newVersionId = reuseOrCreateNewViewVersionId(version);
+
+      if (versions.stream().anyMatch(v -> v.versionId() == newVersionId)) {
+        boolean isNewVersion =
+            lastAddedVersionId != null
+                && changes(MetadataUpdate.AddViewVersion.class)
+                    .anyMatch(added -> added.viewVersion().versionId() == newVersionId);
```

Comment:
The formatting here makes this really hard to read. I think it is better in cases like this to separate expressions and use more variables. Even if you have to split the expression across lines, at least it isn't mixing boolean expression formatting with formatting for long call chains.

```java
boolean isAddedVersion = changes(MetadataUpdate.AddViewVersion.class)
    .anyMatch(added -> added.viewVersion().versionId() == newVersionId);
boolean isNewVersion = lastAddedVersionId != null && isAddedVersion;
```

That said, I don't understand why it works this way. If `addVersionInternal` is called with a view version that already exists, then it should return the existing ID and not set `lastAddedVersionId` because no version was added. That's a simpler way to do what I think you're doing here.

Reply:
should be fixed now with your changes